<!DOCTYPE html>
<html lang="en">
<head>
	<title>Microbializer Overview</title>
	<!--<meta charset="utf-8">-->
	<!--<meta name="viewport" content="width=device-width, initial-scale=1">-->

    <link rel="shortcut icon" type="image/x-icon" href="pics/logo.gif" />
	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
	<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
	<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
	<link rel="stylesheet" href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css">

	<!--files for the select tag-->
	<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/css/bootstrap-select.min.css">
	<script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/bootstrap-select.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/i18n/defaults-*.min.js"></script>

	<!--files for the checkbox tag (toggle)-->
	<script src="https://gitcdn.github.io/bootstrap-toggle/2.2.2/js/bootstrap-toggle.min.js"></script>


	<link rel="stylesheet" href="css/general.css">
	<link rel="stylesheet" href="css/nav.css">

	<!--<link href='https://fonts.googleapis.com/css?family=Kalam' rel='stylesheet'>-->

	<script src="https://code.jquery.com/jquery-1.10.2.js"></script>
	<!--<script> $(function(){-->
		<!--$("head").load("header.html");-->
		<!--$('#loadingMask').fadeOut();});-->
	<!--</script>-->
</head>
<body>
<nav role="navigation" class="navbar navbar-inverse navbar-fixed-top" id="nav">
    <div class="jumbotron" id="jumbo">
        <div class="container">
            <div class="row" id="title-row">
                <div class="col-md-1">
                </div>
                <div class="col-md-10" align="center">
                    <span id="server-title">M1CR0B1AL1Z3R</span>
                    <img src="pics/logo.gif" id="nav_bar_image" class="img-rounded">
                    <br><span id="sub-title">A webserver for analyzing bacterial genomics data. Easily.</span>
					<!--<span id="sub-title">A webserver for analyzing bacterial genomics data</span><br>-->
                </div>
                <!--<div class="col-md-1">-->
                    <!---->
                <!--</div>-->
            </div>
        </div>
    </div>
		<div class="container" id="nav_container">
			<div class="row">
				<div class="col-md-1">
				</div>
				<div class="col-md-10">
					<ul class="nav navbar-nav" id="menu-nav">
						<li>
							<a href="index.html">Home</a>
						</li>
						<li class="active">
							<a href="overview.html">Overview</a>
						</li>
						<li>
							<a href="gallery.html">Gallery</a>
						</li>
						<li>
							<a href="terms.html">Terms of Use</a>
						</li>
						<li>
							<a href="credits.html">Credits</a>
						</li>
						<li style="float: right;">
							<a href="https://english.tau.ac.il/" id="tau-link" target="_blank">
								<img src=pics/TAU_LOGO.png id="tau-logo">
							</a>
						</li>
					</ul>
				</div>
			</div>
		</div>
	</nav>
	<div id="behind-nav-bar">
	</div>
	<div class="container" id="overview_container">
		<div class="row">
			<div class="col-md-1">
			</div>
			<div class="col-md-10" align="justify">
				<h2>Introduction</h2>
				Significant technological advances in the last decade have enabled the use of large-scale sequencing,
				mining, and analysis of bacterial species data in an unprecedented resolution. Such large-scale analyses
				have contributed to the comprehensive characterization of complex microbial dynamics within a microbiome
				[<a href="#ref1">1</a>] and among different strains during disease outbreaks [<a href="#ref2">2</a>].
				Today, such an analysis begins with preliminary steps including genome annotation and orthologs
				detection, sequence alignment, and accurate phylogenetic tree reconstruction. These steps as
				well as additional analysis-specific computations require the use of multiple bioinformatics tools and
				various ad-hoc programming scripts, making the entire process cumbersome, tedious, and prone to errors
				due to manual handling. This has prompted laboratories to implement their own in-house analysis
				pipelines, and thus different analysis applications have begun to emerge [<a href="#ref3">3</a>, <a href="#ref4">4</a>].
				These applications require specific working environments, i.e., operating system, computation power,
				i.e., multi-cores machines, and more than basic technological skills, such as installations and running.
				<br><br>

				<h2>What is M1CR0B1AL1Z3R?</h2>
				M1CR0B1AL1Z3R (pronounced: microbializer) was developed in order to facilitate large scale
				microbial genomics analyses. Our goal was to make M1CR0B1AL1Z3R easily accessible to the scientific
				community and thus to allow the abovementioned computational challenges of analyzing dozens of bacterial
				genomes simultaneously. Such computations are highly demanding, and therefore, M1CR0B1AL1Z3R runs
				background computational processes in parallel on a high-performance computer cluster. Notably, due to
				the high computational demands of these analyses, they cannot run on a simple desktop computer. To this
				end, M1CR0B1AL1Z3R enables the research community to analyze massive genomic data that previously
				necessitated expensive computational infrastructure and expert knowledge. Both visual and textual
				results that are ready for publication or further analysis are given as an output.
				<br><br>

				<h2>Input</h2>
				The input for the webserver is a zipped folder (either <a href="https://support.microsoft.com/en-us/help/14200/windows-compress-uncompress-zip-files" target="_blank">.zip</a> or <a href="https://linhost.info/2012/08/gzip-files-in-windows/" target="_blank">.tar.gz</a>) in which each file is a fasta nucleic acid (<a href="https://en.wikipedia.org/wiki/FASTA_format#FASTA_file" target="_blank">.fna</a>) file containing genomic sequences of a different species (whether fully sequenced or as contigs).

				<br><br>
				<img src="pics/flow.jpg" style="height: 90%; width: 90%;" id="flow">
				<br><br>

				<h2>Output and methodology<br></h2>

				M1CR0B1AL1Z3R implements the following algorithmic steps:
				<ul>
					<li>
					ORF extraction from all genomes using
					Prodigal [<a href="#ref5">5</a>].
					</li>
					<li>
					Homologous gene detection (all against all) and clustering using
					MCL [<a href="#ref6">6</a>, <a href="#ref7">7</a>]. This results in a presence-absence phyletic database
					(a detailed table in which row i is an orthologous group, column j is a bacterial genome and the i,j
					entry contains the member of orthologous family i from genome j, if present.
					</li>
					<li>
					For each orthologous group, M1CR0B1AL1Z3R provides a MAFFT-LINSI [<a href="#ref8">8</a>] multiple
						sequence alignment of all members (the unaligned sequences are provided as well).
					</li>
					<li>
					M1CR0B1AL1Z3R compute the core-genome (those genes that are	shared across all bacterial genomes)
					and provide the core-genome alignment, based on which, the bacterial species tree is often inferred.
					In case where there are duplications (i.e., identical core-genomes), a core-genome alignment containing
					only unique sequences is provided as well.
					</li>
					<li>
					Based on the core-genome alignment a maximum-likelihood
					based phylogenetic tree is reconstructed using RAxML [<a href="#ref9">9</a>] and visualised by
					PhyD3 [<a href="#ref10">10</a>].
					</li>
					<li>
					Several statistics regarding the pan genome (the entire set of
					genomes analyzed) are also provided, e.g., the GC content,
					the distribution of the number of ORFs across the analyzed genomes, and the distribution of the sizes of
					the orthologous groups.
					</li>

				</ul>
				<br><br>


				M1CR0B1AL1Z3R provides the following results:
				<ul>
					<li>
						A curated table of orthologous sets;
					</li>
					<li>
						A fasta file for each orthologous set containing the UNALIGNED sequences of the set members;
					</li>
					<li>
						A fasta file for each orthologous set containing the ALIGNED sequences of the set members;
					</li>
					<li>
						A histogram* of the group sizes frequency.
					</li>
					<li>
						A violin plot* of the ORF counts per genome.
					</li>
					<li>
						A violin plot* of the ORF GC-content per genome.
					</li>
					<li>
						A fasta file containing the inferred core-genome of the species in the dataset;
					</li>
					<li>
						A reliable species phylogeny tree* as an interactive figure;
					</li>
				</ul>
				*Raw data is provided as well
				<br><br>

				Please feel free to try it yourself!<br>
				Click <a href="index_example.html">here</a> submit the provided
				example dataset or go to the <a href="gallery.html">gallery section</a>.
				<br><br>

				<h2>References</h2>

				<li><a href="https://doi.org/10.1038/nrgastro.2017.97" target="_blank" id="ref1">Claesson MJ., et al., <i>Nature Reviews Gastroenterology & Hepatology</i> 14, 585-595 (2017).</a></li>
				<br>
				<li><a href="https://doi.org/10.1371/journal.pgen.1006280" target="_blank" id="ref2">McNally A., et al., <i>Plos Genetics</i> 12, 09 (2016).</a></li>
				<br>
				<li><a href="https://doi.org/10.1093/bioinformatics/btu153" target="_blank" id="ref3">Seemann T., <i>Bioinformatics</i> 30, 2068-2069 (2014).</a></li>
				<br>
				<li><a href="https://doi.org/10.1093/nar/gkx977" target="_blank" id="ref4">Ding W., et al., <i>NAR</i> 46, e5 (2018).</a></li>
				<br>
				<li><a href="https://doi.org/10.1186/1471-2105-11-119" target="_blank" id="ref5">Hyatt D., et al., <i>BMC Bioinformatics</i> 11, 119 (2010).</a></li>
				<br>
				<li><a href="https://doi.org/10.1016/S0022-2836(05)80360-2" target="_blank" id="ref6">Altschul SF., et al., <i>J Mol Biol</i> 215, 403-410 (1990).</a></li>
				<br>
				<li><a href="https://dspace.library.uu.nl/bitstream/handle/1874/848/full.pdf?sequence=1&isAllowed=y" target="_blank" id="ref7">Van Dongen SM., <i>Thesis</i> (2000).</a></li>
				<br>
				<li><a href="https://doi.org/10.1093/molbev/mst010" target="_blank" id="ref8">Katoh K. & Standley DM., <i>Mol Biol Evol</i> 30, 772-780 (2013)</a></li>
				<br>
				<li><a href="https://doi.org/10.1093/bioinformatics/btu033" target="_blank" id="ref9">Stamatakis A., <i>Bioinformatics</i> 30, 1312-1313 (2014)</a></li>
				<br>
				<li><a href="https://doi.org/10.1093/bioinformatics/btx324" target="_blank" id="ref10">Kreft L., et al., <i>Bioinformatics</i> 33, 2946-2947 (2017)</a></li>
				<br><br>
			</div>
		</div>
	</div>
</body>
</html>







